src/decoder.rs

```diff
@@ -101,14 +101,195 @@ impl Decoder {
+                    if neck_inst.rd() != neck_inst.rs1()
+                        || neck_inst.rs1() != next_inst.rs1()
+                        || neck_inst.rs2() == head_inst.rs1()
+                        || neck_inst.rs2() == head_inst.rs2()
+                    {
+                        return Ok(None);
+                    }
+                    let body_instruction = self.decode_raw(
+                        memory,
+                        pc + head_size as u64 + next_size as u64 + neck_size as u64,
+                    )?;
+                    let body_opcode = extract_opcode(body_instruction);
+                    if body_opcode != insts::OP_SLTU {
+                        return Ok(None);
+                    }
+                    let body_inst = Rtype(body_instruction);
+                    let body_size = instruction_length(body_instruction);
+                    if body_inst.rd() != body_inst.rs2()
+                        || body_inst.rs2() != neck_inst.rs2()
+                        || body_inst.rs1() != neck_inst.rs1()
+                    {
+                        return Ok(None);
+                    }
+                    let tail_instruction = self.decode_raw(
+                        memory,
+                        pc + head_size as u64
+                            + next_size as u64
+                            + neck_size as u64
+                            + body_size as u64,
+                    )?;
+                    let tail_opcode = extract_opcode(tail_instruction);
+                    if tail_opcode != insts::OP_OR {
+                        return Ok(None);
+                    }
+                    let tail_inst = Rtype(tail_instruction);
+                    let tail_size = instruction_length(tail_instruction);
+                    if tail_inst.rd() != tail_inst.rs1()
+                        || tail_inst.rs1() != head_inst.rs2()
+                        || tail_inst.rs2() != body_inst.rs2()
+                    {
+                        return Ok(None);
+                    }
+                    let fuze_inst = Rtype::new(
+                        insts::OP_ADC,
+                        head_inst.rd(),
+                        next_inst.rd(),
+                        body_inst.rd(),
+                    );
+                    let fuze_size = head_size + next_size + neck_size + body_size + tail_size;
+                    Ok(Some(set_instruction_length_n(fuze_inst.0, fuze_size)))
+                };
+                if let Ok(Some(i)) = rule_adc() {
+                    Ok(i)
+                } else {
+                    Ok(head_instruction)
+                }
+            }
+            insts::OP_SUB => {
+                let mut rule_sbb = || -> Result<Option<Instruction>, Error> {
+                    let head_inst = Rtype(head_instruction);
+                    let head_size = instruction_length(head_instruction);
+                    if head_inst.rs1() == ZERO || head_inst.rs2() == ZERO {
```

Comment:
For adc, we were checking rd of `head_inst` is not ZERO, why this is missing here?

Reply:
I think it might be better to have a unified check, where we are checking all 3 registers used are not ZERO.

Reply:
The next codes will check head_inst.rd() == head_inst.rs2()

Reply:
Yep but we are relying an inherit rule that is quite prone to errors, I do suggest the following changes:

* When we were decoding instructions, we only check for the equality of certain register IDs, such as `head_inst.rd() == head_inst.rs2()`
* When we have decoded all the instructions, and are ready to assemble the mop adc/sbb instruction, we add a ZERO register check for all used registers by the mop instruction

I think this way we can have better code readability, and I do suspect the generated optimized code, will be the same level of performance.

Reply:
Yes, this way is better